File: hdf-diff/src/sbom.ts
/**
 * SBOM comparison for the HDF diff engine.
 *
 * Compares two SBOM documents (CycloneDX or SPDX JSON) and produces
 * package-level diffs: added, removed, updated, unchanged packages.
 *
 * Format auto-detection:
 * - CycloneDX: has `bomFormat: "CycloneDX"`, packages in `components[]`
 * - SPDX: has `spdxVersion`, packages in `packages[]`
 */

/**
 * The comparison result for a single package between two SBOMs.
 */
export interface PackageDiff {
  purl: string;
  name: string;
  state: 'added' | 'removed' | 'updated' | 'unchanged';
  oldVersion?: string;
  newVersion?: string;
  licenses?: string[];
}

/**
 * The complete result of comparing two SBOM documents.
 */
export interface SbomDiffResult {
  packageDiffs: PackageDiff[];
  added: number;
  removed: number;
  updated: number;
  unchanged: number;
}

/**
 * Internal representation of a parsed package from any SBOM format.
 */
interface ParsedPackage {
  purl: string;
  name: string;
  version: string;
  licenses: string[];
}

/**
 * Compare two SBOM documents (CycloneDX or SPDX JSON strings) and return
 * package-level diffs.
 *
 * @param oldJson - JSON string of the old SBOM document
 * @param newJson - JSON string of the new SBOM document
 * @returns Structured diff result with per-package state and aggregate counts
 * @throws Error if either input is not valid JSON or not a recognized SBOM format
 */
export function diffSboms(oldJson: string, newJson: string): SbomDiffResult {
  const oldParsed = JSON.parse(oldJson) as Record<string, unknown>;
  const newParsed = JSON.parse(newJson) as Record<string, unknown>;

  const oldPackages = extractPackages(oldParsed);
  const newPackages = extractPackages(newParsed);

  const oldMap = buildPackageMap(oldPackages);
  const newMap = buildPackageMap(newPackages);

  const diffs: PackageDiff[] = [];
  const seen = new Set<string>();

  // Check old packages against new
  for (const [key, oldPkg] of oldMap) {
    seen.add(key);
    const newPkg = newMap.get(key);
    if (newPkg) {
      if (oldPkg.version !== newPkg.version) {
        diffs.push({
          purl: newPkg.purl || newPkg.name,
          name: newPkg.name,
          state: 'updated',
          oldVersion: oldPkg.version,
          newVersion: newPkg.version,
          licenses: newPkg.licenses.length > 0 ? newPkg.licenses : undefined,
        });
      } else {
        diffs.push({
          purl: oldPkg.purl || oldPkg.name,
          name: oldPkg.name,
          state: 'unchanged',
        });
      }
    } else {
      diffs.push({
        purl: oldPkg.purl || oldPkg.name,
        name: oldPkg.name,
        state: 'removed',
        oldVersion: oldPkg.version,
      });
    }
  }

  // Check for added packages
  for (const [key, newPkg] of newMap) {
    if (!seen.has(key)) {
      diffs.push({
        purl: newPkg.purl || newPkg.name,
        name: newPkg.name,
        state: 'added',
        newVersion: newPkg.version,
        licenses: newPkg.licenses.length > 0 ? newPkg.licenses : undefined,
      });
    }
  }

  // Sort by name for deterministic output
  diffs.sort((a, b) => a.name.localeCompare(b.name));

  // Count states
  let added = 0;
  let removed = 0;
  let updated = 0;
  let unchanged = 0;
  for (const d of diffs) {
    switch (d.state) {
      case 'added': added++; break;
      case 'removed': removed++; break;
      case 'updated': updated++; break;
      case 'unchanged': unchanged++; break;
    }
  }

  return { packageDiffs: diffs, added, removed, updated, unchanged };
}

/**
 * Extract packages from an SBOM document, auto-detecting the format.
 */
function extractPackages(doc: Record<string, unknown>): ParsedPackage[] {
  if (doc['bomFormat'] === 'CycloneDX') {
    return extractCycloneDXPackages(doc);
  }
  if (typeof doc['spdxVersion'] === 'string') {
    return extractSPDXPackages(doc);
  }
  throw new Error(
    'Unrecognized SBOM format: expected CycloneDX (bomFormat) or SPDX (spdxVersion)',
  );
}

/**
 * Extract packages from a CycloneDX document.
 * Reads components[] with purl, name, version fields.
 */
function extractCycloneDXPackages(doc: Record<string, unknown>): ParsedPackage[] {
  const components = doc['components'] as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(components)) {
    return [];
  }

  return components.map((comp): ParsedPackage => {
    const licenses = extractCycloneDXLicenses(comp);
    return {
      purl: (comp['purl'] as string) ?? '',
      name: (comp['name'] as string) ?? '',
      version: (comp['version'] as string) ?? '',
      licenses,
    };
  });
}

/**
 * Extract license strings from a CycloneDX component.
 * CycloneDX licenses can be in `licenses[].license.id` or `licenses[].license.name`.
 */
function extractCycloneDXLicenses(comp: Record<string, unknown>): string[] {
  const licensesArr = comp['licenses'] as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(licensesArr)) {
    return [];
  }
  const result: string[] = [];
  for (const entry of licensesArr) {
    const license = entry['license'] as Record<string, unknown> | undefined;
    if (license) {
      const id = license['id'] as string | undefined;
      const name = license['name'] as string | undefined;
      if (id) {
        result.push(id);
      } else if (name) {
        result.push(name);
      }
    }
  }
  return result;
}

/**
 * Extract packages from an SPDX document.
 * Reads packages[] with name, versionInfo, and externalRefs (for PURL).
 */
function extractSPDXPackages(doc: Record<string, unknown>): ParsedPackage[] {
  const packages = doc['packages'] as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(packages)) {
    return [];
  }

  return packages.map((pkg): ParsedPackage => {
    const purl = extractSPDXPurl(pkg);
    const license = (pkg['licenseConcluded'] as string) ?? (pkg['licenseDeclared'] as string) ?? '';
    return {
      purl,
      name: (pkg['name'] as string) ?? '',
      version: (pkg['versionInfo'] as string) ?? '',
      licenses: license && license !== 'NOASSERTION' ? [license] : [],
    };
  });
}

/**
 * Extract a PURL from an SPDX package's externalRefs array.
 */
function extractSPDXPurl(pkg: Record<string, unknown>): string {
  const refs = pkg['externalRefs'] as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(refs)) {
    return '';
  }
  for (const ref of refs) {
    const refType = ref['referenceType'] as string | undefined;
    if (refType === 'purl') {
      return (ref['referenceLocator'] as string) ?? '';
    }
  }
  return '';
}

/**
 * Build a lookup map of packages indexed by name (version-independent key).
 */
function buildPackageMap(packages: ParsedPackage[]): Map<string, ParsedPackage> {
  const map = new Map<string, ParsedPackage>();
  for (const pkg of packages) {
    const key = pkg.name || pkg.purl;
    if (key) {
      map.set(key, pkg);
    }
  }
  return map;
}

File: hdf-diff/go/sbom.go
// Package diff provides SBOM comparison functionality for the HDF diff engine.
// It compares two SBOM documents (CycloneDX or SPDX JSON) and produces
// package-level diffs showing added, removed, updated, and unchanged packages.
//
// Uses protobom for format-agnostic SBOM parsing.
package diff

import (
	"bytes"
	"fmt"
	"sort"

	"github.com/protobom/protobom/pkg/reader"
	"github.com/protobom/protobom/pkg/sbom"
)

// PackageDiff represents the comparison result for a single package between two SBOMs.
type PackageDiff struct {
	Purl       string   `json:"purl"`
	Name       string   `json:"name"`
	State      string   `json:"state"` // added, removed, updated, unchanged
	OldVersion string   `json:"oldVersion,omitempty"`
	NewVersion string   `json:"newVersion,omitempty"`
	Licenses   []string `json:"licenses,omitempty"`
}

// DiffResult holds the complete result of comparing two SBOM documents.
//
//nolint:revive // matches TypeScript export name
type DiffResult struct {
	PackageDiffs []PackageDiff `json:"packageDiffs"`
	Added        int           `json:"added"`
	Removed      int           `json:"removed"`
	Updated      int           `json:"updated"`
	Unchanged    int           `json:"unchanged"`
}

// DiffSBOMs compares two SBOM files (CycloneDX or SPDX JSON) and returns
// package-level diffs. Uses protobom for format-agnostic parsing.
//
//nolint:revive // matches TypeScript export name
func DiffSBOMs(oldData, newData []byte) (*DiffResult, error) {
	r := reader.New()

	oldDoc, err := r.ParseStream(bytes.NewReader(oldData))
	if err != nil {
		return nil, fmt.Errorf("failed to parse old SBOM: %w", err)
	}

	newDoc, err := r.ParseStream(bytes.NewReader(newData))
	if err != nil {
		return nil, fmt.Errorf("failed to parse new SBOM: %w", err)
	}

	oldNodes := extractNodes(oldDoc)
	newNodes := extractNodes(newDoc)

	oldMap := buildNodeMap(oldNodes)
	newMap := buildNodeMap(newNodes)

	var diffs []PackageDiff
	seen := make(map[string]bool)

	// Check old packages against new
	for key, oldNode := range oldMap {
		seen[key] = true
		if newNode, exists := newMap[key]; exists {
			if oldNode.Version != newNode.Version {
				diffs = append(diffs, PackageDiff{
					Purl:       nodeKey(newNode),
					Name:       newNode.Name,
					State:      "updated",
					OldVersion: oldNode.Version,
					NewVersion: newNode.Version,
					Licenses:   nonEmptyLicenses(newNode.Licenses),
				})
			} else {
				diffs = append(diffs, PackageDiff{
					Purl:  nodeKey(oldNode),
					Name:  oldNode.Name,
					State: "unchanged",
				})
			}
		} else {
			diffs = append(diffs, PackageDiff{
				Purl:       nodeKey(oldNode),
				Name:       oldNode.Name,
				State:      "removed",
				OldVersion: oldNode.Version,
			})
		}
	}

	// Check for added packages
	for key, newNode := range newMap {
		if !seen[key] {
			diffs = append(diffs, PackageDiff{
				Purl:       nodeKey(newNode),
				Name:       newNode.Name,
				State:      "added",
				NewVersion: newNode.Version,
				Licenses:   nonEmptyLicenses(newNode.Licenses),
			})
		}
	}

	// Sort by name for deterministic output
	sort.Slice(diffs, func(i, j int) bool { return diffs[i].Name < diffs[j].Name })

	result := &DiffResult{PackageDiffs: diffs}
	for _, d := range diffs {
		switch d.State {
		case "added":
			result.Added++
		case "removed":
			result.Removed++
		case "updated":
			result.Updated++
		case "unchanged":
			result.Unchanged++
		}
	}

	return result, nil
}

// extractNodes safely retrieves the node list from a protobom document.
func extractNodes(doc *sbom.Document) []*sbom.Node {
	nl := doc.GetNodeList()
	if nl == nil {
		return nil
	}
	return nl.GetNodes()
}

// buildNodeMap indexes nodes by their canonical key (PURL without version, or name).
// When multiple nodes share the same key, the last one wins — this mirrors
// how package managers resolve duplicate entries.
func buildNodeMap(nodes []*sbom.Node) map[string]*sbom.Node {
	m := make(map[string]*sbom.Node, len(nodes))
	for _, n := range nodes {
		key := matchKey(n)
		if key != "" {
			m[key] = n
		}
	}
	return m
}

// matchKey returns a version-independent key for matching a node across SBOMs.
// Uses the PURL name portion (stripping version) if available, otherwise the node name.
func matchKey(n *sbom.Node) string {
	// Use node name as the match key. The PURL contains the version, so using
	// the raw name gives us a version-independent match key.
	if n.Name != "" {
		return n.Name
	}
	// Fallback to PURL string if name is empty
	purl := string(n.Purl())
	if purl != "" {
		return purl
	}
	return ""
}

// nodeKey returns the best identifier for a node: PURL if available, else name.
func nodeKey(n *sbom.Node) string {
	purl := string(n.Purl())
	if purl != "" {
		return purl
	}
	return n.Name
}

// nonEmptyLicenses returns the license slice only if it has entries, nil otherwise.
// This keeps JSON output clean by omitting empty arrays.
func nonEmptyLicenses(licenses []string) []string {
	if len(licenses) == 0 {
		return nil
	}
	return licenses
}
